File: csharpcore/TellDontAskKata/TellDontAskKata.Main/UseCase/OrderApprovalUseCase.cs
using TellDontAskKata.Main.Domain;
using TellDontAskKata.Main.Repository;

namespace TellDontAskKata.Main.UseCase
{
    public class OrderApprovalUseCase
    {
        private readonly IOrderRepository _orderRepository;
        public OrderApprovalUseCase(IOrderRepository orderRepository)
        {
            _orderRepository = orderRepository;
        }

        public void Run(OrderApprovalRequest request)
        {
            var order = _orderRepository.GetById(request.OrderId);

            if (order.Status == OrderStatus.Shipped)
            {
                throw new ShippedOrdersCannotBeChangedException();
            }

            if (request.Approved && order.Status == OrderStatus.Rejected)
            {
                throw new RejectedOrderCannotBeApprovedException();
            }

            if (!request.Approved && order.Status == OrderStatus.Approved)
            {
                throw new ApprovedOrderCannotBeRejectedException();
            }

            order.Status = request.Approved ? OrderStatus.Approved : OrderStatus.Rejected;
            _orderRepository.Save(order);
        }
    }
}

File: scala/src/main/scala/ordershipping/usecase/OrderApprovalUseCase.scala
package ordershipping.usecase

import ordershipping.domain.OrderStatus
import ordershipping.repository.OrderRepository

class OrderApprovalUseCase(val orderRepository: OrderRepository) {
  def run(request: OrderApprovalRequest): Unit = {
    orderRepository
      .getById(request.orderId)
      .foreach(order => {
        if (order.status == OrderStatus.Shipped)
          throw new ShippedOrdersCannotBeChangedException
        if (request.approved && order.status == OrderStatus.Rejected)
          throw new RejectedOrderCannotBeApprovedException
        if (!request.approved && order.status == OrderStatus.Approved)
          throw new ApprovedOrderCannotBeRejectedException

        order.status =
          if (request.approved) OrderStatus.Approved else OrderStatus.Rejected
        orderRepository.save(order)
      })
  }
}
